pkg/cloudprovider/azure.go

```diff
@@ -562,26 +581,65 @@ func (a *Azure) getAuthorizer(env azureapi.Environment, clientID, clientSecret,
 		cred azcore.TokenCredential
 		err  error
 	)
-	if a.azureWorkloadIdentityEnabled && strings.TrimSpace(clientSecret) == "" {
-		options := azidentity.WorkloadIdentityCredentialOptions{
+
+	// MSI Override for ARO HCP
+	msi := os.Getenv("AZURE_MSI_AUTHENTICATION")
+	if msi == "true" {
+		options := azidentity.ManagedIdentityCredentialOptions{
 			ClientOptions: azcore.ClientOptions{
 				Cloud: cloudConfig,
 			},
-			ClientID:      clientID,
-			TenantID:      tenantID,
-			TokenFilePath: tokenFile,
 		}
-		cred, err = azidentity.NewWorkloadIdentityCredential(&options)
+
+		var err error
+		cred, err = azidentity.NewManagedIdentityCredential(&options)
 		if err != nil {
 			return nil, err
 		}
+	} else if strings.TrimSpace(cfg.clientSecret) == "" {
+		if a.azureWorkloadIdentityEnabled && strings.TrimSpace(cfg.tokenFile) != "" {
+			klog.Infof("Using workload identity authentication")
+			if cfg.clientID == "" || cfg.tenantID == "" {
+				return nil, fmt.Errorf("clientID and tenantID are required in workload identity authentication")
+			}
+			options := azidentity.WorkloadIdentityCredentialOptions{
+				ClientOptions: azcore.ClientOptions{
+					Cloud: cloudConfig,
+				},
+				ClientID:      cfg.clientID,
+				TenantID:      cfg.tenantID,
+				TokenFilePath: cfg.tokenFile,
+			}
+			cred, err = azidentity.NewWorkloadIdentityCredential(&options)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			klog.Infof("Using managed identity authentication")
+			options := azidentity.ManagedIdentityCredentialOptions{
+				ClientOptions: azcore.ClientOptions{
+					Cloud: cloudConfig,
+				},
+			}
+			if cfg.clientID != "" {
+				options.ID = azidentity.ClientID(cfg.clientID)
+			}
+			cred, err = azidentity.NewManagedIdentityCredential(&options)
+			if err != nil {
+				return nil, err
+			}
+		}
```

Comment:
This is no longer necessary with the new env var check above.

Reply:
I think we should remove it, this code is not reachable with AZURE_FEDERATED_TOKEN_FILE being defaulted to a constant path.